matchzoo/engine/base_transformer.py

```diff
@@ -0,0 +1,296 @@
+"""Contains the base Transformer class for  all transformers."""
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+
+import os
+import abc
+import typing
+import pickle
+import re
+import six
+from collections import Mapping
+
+from matchzoo import engine
+
+
+class BaseTransformer(abc.ABC):
+    """Base data transformer for MatchZoo.
+
+    # Methods:
+    ==========
+    fit_transform(X, y=None, **fit_params)
+
+    fit(X, y=None)
+
+    transform(X, copy=True)
+
+    """
+
+    _white_spaces = re.compile(r"\s\s+")
+
+    def __init__(self, params: typing.Optional[engine.TransformerParams]
+                 = None):
+        """:class:`BaseTransformer` constructor."""
+        self._params = params or self.get_default_params()
+
+    @classmethod
+    def get_default_params(cls) -> engine.TransformerParams:
+        """Return the default parameters for transformer."""
+        return engine.TransformerParams()
+
+    def guess_and_fill_missing_params(self):
+        """
+        Guess and fill missing parameters in :attribute:`params`.
+
+        Note: likely to be moved to a higher level API in the future.
+        """
+        if self._params['name'] is None:
```

Comment:
`if X` == `if X is None`

Reply:
Object boolean checking in params is a special case, where `None` has a special meaning of "not filled yet", and anything else means "already filled". For example, a param like "dropout" could be set to 0, which is reasonable number but its boolean is `False`. I solved this problem locally by making a single param a class, and overrides its `__bool__`, and we can go from there.